### PR TITLE
Enable Entity Linking and Entity Based Search

### PR DESCRIPTION
**Pull request by ankur-gos**
This pull request contains code to:
- Link entities to [UMLS](https://www.nlm.nih.gov/research/umls/index.html) using [scispacy](https://github.com/allenai/scispacy)
- Reorganize elastic indices to take entity-object relationship into account
- Enable entity based search
    - Using ElasticSearch to link on the name, return all relevant contexts (with class conditioning, scoring, etc) as well as the linked entities types and descriptions
- Add CLI scripts to ingest/delete indices into elastic instead of having the script in the docker-compose

The main code for linking resides in ingestion/process/linking, while the ElasticRetriever has relevant updates to enable entity search.

### CODE REVIEW

**Review by iross — COMMENTED** [review]
Ugh, sorry. I thought I was adding comments to the PR but was adding them to a review that I then didn't 'submit'.

Mostly looks good, but I have questions on the performance both of the ingest + the query. The ingest I really would like to do via bulk transaction.

File: cosmos/retrieval/retrieval/elastic_retriever.py [hunk 1]
```diff
@@ -11,7 +11,89 @@
 logger.setLevel(logging.DEBUG)
 
 
-class Object(Document):
+class EntityObjectIndex(Document):
+    """
+    We need to put entities and the objects associated on the same shard, so we are going have them inherit
+    from this index class
+    """
+    entity_object = Join(relations={"entity": "object"})
```
[the remark this answers is not in the document]
Comment (ankur-gos) [reply cosmos/retrieval/retrieval/elastic_retriever.py:19]: I would expect that the performance is OK, given that the objects are denormalized and duplicated across entities. We can do perform a profile in the future.

File: cosmos/retrieval/retrieval/elastic_retriever.py [hunk 2]
```diff
@@ -11,7 +11,89 @@
...
+    @classmethod
+    def _matches(cls, hit):
+        """ Use Entity class for parent documents """
+        return hit["_source"]["entity_object"] == "entity"
+
+    @classmethod
+    def search(cls, **kwargs):
+        return cls._index.search(**kwargs).filter("term", entity_object="entity")
+
+    def add_object(self, cls, dataset_id, content, header_content, area, detect_score, postprocess_score, pdf_name, img_path=None, commit=True):
```
[the remark this answers is not in the document]
Comment (ankur-gos) [reply cosmos/retrieval/retrieval/elastic_retriever.py:50]: This is important. I'll make the bulk change.

File: cosmos/retrieval/retrieval/elastic_retriever.py [hunk 3]
```diff
@@ -11,7 +11,89 @@
...
+            dataset_id=dataset_id,
+            content=content,
+            header_content=header_content,
+            area=area,
+            detect_score=detect_score,
+            postprocess_score=postprocess_score,
+            pdf_name=pdf_name,
+            img_path=img_path
+        )
+        if commit:
```
[the remark this answers is not in the document]
Comment (ankur-gos) [reply cosmos/retrieval/retrieval/elastic_retriever.py:69]: specifically if you want to do things like bulk inserts.

File: cli/link_parquets.sh [hunk 4]
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
+python3.8 -m ingest.process.entity_linking.link --input-path /input --output-path /output/ \
+    --cluster tcp://scheduluer:8786 --dataset-id documents
```
Comment (ilmcconnell) [inline cli/link_parquets.sh:3]: Spelling on scheduluer

File: cosmos/retrieval/retrieval/elastic_retriever.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (iross) [reply cosmos/retrieval/retrieval/elastic_retriever.py:58]: Added in acda80e

File: cli/link_parquets.sh (hunk 4 above)
Comment (iross) [reply cli/link_parquets.sh]: It actually was a big deal ;)  Fixed in 0463b8d

**Comment by iross** [comment]
My big issue with this implementation is the duplication of the objects for each occurrence of an entity. It seems wasteful, but _ok_ in the table/figure caption cases, since there's a low number of entities in those. But sections ends up being pretty bloated and equations even worse. In a test of 1,000 docs with 1158 equations, I'm seeing over **500,000** objects with cls:Equation in eo-site, verified by looking at the average number of  linked entities for the equation dataframe:

```
>>> q['ents_linked'].apply(len).mean()
463.2262521588946
```

Worse, each equation is associated with its page contents (https://github.com/UW-COSMOS/Cosmos/blob/d60cfb18196052d2155562957aaf317dae62b502/cosmos/ingestion/ingest/process/aggregation/aggregate.py#L229). So we're getting both huge numbers of entities linked to each equation _and_ we're saving 463 copies of these pages' contents.. The end result is a 20GB index for a 1,000 document set. So we'd be looking at a ~2TB index for the covid set.

I'll experiment with some alternatives..

**Comment by ankur-gos** [comment]
This is intentional. Having it be denormalized like this makes it faster to query (we don’t need to do joins), at the cost of space. The alternative is a normalized entity/object table, which will be really slow to join on an online setting like we have.

We have a normalized join happening between pdfs and objects, if you want to implement a normalized version.

**Review by iross — APPROVED** [review]
I think the ES/API aspect still needs some consideration. It definitely works in the current state, I'm just still unconvinced about the space-vs-performance tradeoffs. The pipeline aspects look good, as usual, so I'll go ahead and merge.

**Pull request merged**